File: sdk-core/src/main/java/dev/oppex/sdk/exception/IncidentException.java
package dev.oppex.sdk.exception;

/** Checked failure raised by synchronous incident delivery. */
public class IncidentException extends Exception {
    private static final long serialVersionUID = 1L;

    private final int statusCode;
    private final boolean retryable;

    public IncidentException(String message) {
        this(message, null, -1, false);
    }

    public IncidentException(String message, Throwable cause) {
        this(message, cause, -1, false);
    }

    public IncidentException(String message, int statusCode, boolean retryable) {
        this(message, null, statusCode, retryable);
    }

    public IncidentException(String message, Throwable cause, int statusCode, boolean retryable) {
        super(message, cause);
        this.statusCode = statusCode;
        this.retryable = retryable;
    }

    /** Returns the HTTP status, or -1 when no HTTP response was received. */
    public int getStatusCode() {
        return statusCode;
    }

    /** Returns whether the underlying failure is eligible for retry. */
    public boolean isRetryable() {
        return retryable;
    }
}

File: dotnet/src/Oppex.Integration.Sdk/IncidentException.cs
namespace Oppex.Integration.Sdk;

/// <summary>Thrown when incident delivery was attempted and failed.</summary>
/// <remarks>
/// Validation failures throw <see cref="ArgumentException"/> and a post made
/// after disposal throws <see cref="ObjectDisposedException"/>; neither reaches
/// the network, so neither is an <see cref="IncidentException"/>.
/// </remarks>
public sealed class IncidentException : Exception
{
    /// <summary>
    /// The <see cref="StatusCode"/> reported when delivery failed before any
    /// HTTP status line was received.
    /// </summary>
    public const int NoStatusCode = -1;

    /// <summary>Creates a failure that never reached a status line.</summary>
    public IncidentException(string message)
        : this(message, NoStatusCode, retryable: false, innerException: null)
    {
    }

    /// <summary>Creates a failure that never reached a status line.</summary>
    public IncidentException(string message, Exception? innerException)
        : this(message, NoStatusCode, retryable: false, innerException)
    {
    }

    /// <summary>Creates a failure carrying its HTTP status and retry decision.</summary>
    public IncidentException(string message, int statusCode, bool retryable, Exception? innerException = null)
        : base(message, innerException)
    {
        StatusCode = statusCode;
        Retryable = retryable;
    }

    /// <summary>
    /// The HTTP status, or <see cref="NoStatusCode"/> when no response arrived.
    /// </summary>
    public int StatusCode { get; }

    /// <summary>Whether the failure was eligible for retry.</summary>
    public bool Retryable { get; }

    /// <summary>Whether an HTTP status was received at all.</summary>
    public bool HasHttpStatus => StatusCode != NoStatusCode;
}
